systemd service hardening

gangaram-tii opened the pull request:
<!--
    Copyright 2023 TII (SSRC) and the Ghaf contributors
    SPDX-License-Identifier: CC-BY-SA-4.0
-->

## Description of changes
Systemd service hardening:
- Hardened configuration of different service
- By default hardened configuration enabled in all VMs
- Two services `Network manager`, `sshd`, and `user` are hardened only for release build
- Any service can be excluded from hardening

Services are hardened using the following mechanisms wherever applicable:
- Service Sandboxing
- Filesystem Access Control
- Kernel Parameter Protection
- Network Access Restrictions
- Secure Execution Policies
- Kernel Log Access Control

For each hardened service, the exposure level is evaluated using the `systemd-analyze security` command. Based on their functionality, services are configured to have restricted access to minimize potential vulnerabilities.

The configs used for hardening are listed in the doc:

[Systemd hardening options](https://github.com/gangaram-tii/ghaf/blob/__alive-systemd-hardened-configs/modules/common/systemd/hardened-configs/config-options.md)

<!--
Summary of the proposed changes in the PR description in your own words. For dependency updates, please link to the changelog.
-->

## Checklist for things done

<!-- Please check, [X], to all that applies. Leave [ ] if an item does not apply but you have considered the check list item. Note that all of these are not hard requirements. They serve information to reviewers. When you fill the checklist, you indicate to reviewers you appreciate their work. -->

- [x] Summary of the proposed changes in the PR description
- [x] More detailed description in the commit message(s)
- [x] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [x] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [x] Test procedure described (or includes tests). Select one or more:
  - [x] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [x] Author has run `nix flake check --accept-flake-config` and it passes
- [x] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [x] Author has added reviewers and removed PR draft status

<!-- Additional description of omitted [ ] items if not obvious. -->

## Testing

Tested the functionality of each application on a Lenovo X1. The system logs from each virtual machine (VM) were compared with their corresponding non-hardened versions. No new warnings or errors were observed as a result of hardening.

Additionally, tested configurations in a dummy service that runs the `sandcat` agent from Caldera (assumption of compromised service). The agent was found to be restricted according to the hardened service configurations. In contrast, when the same agent ran in a service without hardened configurations, it was able to carry out attacks."

Requires regressive testing. 

<!--
How this was tested by the author? How is this supposed to be tested
by people doing system testing?
-->

gangaram-tii commented:
recheck

leivos-unikie reviewed (approved):
Tested on Lenovo-X1-gen11 (boot from USB SSD)
- ci-test-automation passed, no changes in performance results
- Checked functionality of all Apps

gangaram-tii commented:
> Check license

Updated the doc.

brianmcgillion commented:
just to be clear it needs to be tested on all platforms as it is a global change

mbssrc commented:
Please test this also on `x86_64-generic/vm` and  `agx/nx` as it enables hardening for `microvm-host` and `netvm`, which are present in these targets.

milva-unikie commented:
**Orin-AGX (nvidia-jetson-orin-agx-debug)**
- Ssh connection to net-vm does not work.
```
[ghaf@ghaf-host:~]$ sudo ssh ghaf@192.168.101.1
ssh: connect to host 192.168.101.1 port 22: No route to host

[ghaf@ghaf-host:~]$ sudo ssh ghaf@net-vm.ghaf
ssh: Could not resolve hostname net-vm.ghaf: Name or service not known
```
- Serial connection works with `sudo minicom -D /dev/ttyACM1` and shows that net-vm is active but does not have an ip address. 
- All apps work.


**Orin-AGX Cross (nvidia-jetson-orin-agx-debug-from-x86_64)**
- Same as native build, net-vm active but can not connect with ssh.
- All apps work.


**Orin-NX (nvidia-jetson-orin-nx-debug)**
- Same as AGX, can not connect with ssh.
- All apps work.

milva-unikie commented:
**NUC (generic-x86_64-debug)**
- Test-automation passes
- Ssh to net-vm works
- All apps work

**VM-debug (nix run .#packages.x86_64-linux.vm-debug)**
- Ssh to net-vm works
- All apps work

gangaram-tii commented:
> nvidia-jetson-orin-nx-debug

Can you please share the journalctl log from host and net-vm?

gangaram-tii commented:
Updated the PR in forced commit.
Did following changes:
- Moved systemd hardening document under docs.
- Updated the systemd-udev-trigger service config to fix the ssh issue:

```
diff --git a/modules/common/systemd/hardened-configs/common/systemd-udev-trigger.nix b/modules/common/systemd/hardened-configs/common/systemd-udev-trigger.nix
index aa0d82f..eca08d4 100644
--- a/modules/common/systemd/hardened-configs/common/systemd-udev-trigger.nix
+++ b/modules/common/systemd/hardened-configs/common/systemd-udev-trigger.nix
@@ -6,7 +6,7 @@
   # Networking #
   ##############
 
-  PrivateNetwork = true;
+  # PrivateNetwork = true;
   # IPAccounting=yes
   IPAddressDeny = "any";
   RestrictAddressFamilies = [
@@ -66,6 +66,7 @@
   ProtectClock = true;
   ProtectControlGroups = true;
   RestrictNamespaces = true;

   /*
     RestrictNamespaces=[
    #"~user"
```

gangaram-tii commented:
> Maybe hardened configs could be cleaned up? So important comments are not buried between commented boilerplate code.

I left the unused configs commented for following reasons:
- Uniformity across different services
- Easy to compare two service configs
- Further tuning and debugging is easy as all focused configs are already there, we need noy to search any where.

gangaram-tii commented:
@leivos-unikie Can you please test the updated PR on Nvidia platform?

milva-unikie commented:
**Orin-NX (nvidia-jetson-orin-nx-debug)**
- Test-automation passes, can connect to net-vm.
- All apps work.

We will test AGX tomorrow. I had some problems in flashing but did not have time to  take a closer look yet. They might have nothing to do with this pr.

johannarautanen commented:
**Orin-AGX (nvidia-jetson-orin-agx-debug-flash-script)**

- Test-automation passes, can connect to net-vm.
- All apps work.

milva-unikie commented:
**Orin-AGX Cross (nvidia-jetson-orin-agx-debug-from-x86_64-flash-script)**
- Test-automation passes, can connect to net-vm.
- All apps work.

I was also today able to flash the native AGX image, so that was just a temporary issue.

gangaram-tii commented:
Can anyone who has rights, re-trigger ghaf-pre-merge-pipeline build? The failure is intermittent, it is not related to the PR.
```
11:51:56         last 10 log lines:
11:51:56         > + mountOptions=('umask=0077')
11:51:56         > + declare -a mountOptions
11:51:56         > + mountpoint=/boot
11:51:56         > + type=filesystem
11:51:56         > + findmnt /dev/disk/by-partlabel/disk-disk1-ESP /mnt/boot
11:51:56         > + mount /dev/disk/by-partlabel/disk-disk1-ESP /mnt/boot -t vfat -o umask=0077 -o X-mount.mkdir
11:51:56         > + rm -rf /tmp/tmp.Gz19IEgK9U
11:51:56         > warning: the group 'nixbld' specified in 'build-users-group' does not exist
11:51:56         > cp: cannot access '/nix/store/w7hmmw35p0ylsgdy4a33nmrl4qyzii42-duktape-2.7.0': Cannot allocate memory
11:51:56         > [  832.840794] reboot: Power down
11:51:56         For full logs, run 'nix log /nix/store/97963q2q4hxjzi02qsi2j69y2zic2y0x-ghaf-host-disko-images.drv'.
11:51:56  error: builder for '/nix/store/97963q2q4hxjzi02qsi2j69y2zic2y0x-ghaf-host-disko-images.drv' failed with exit code 1;
11:51:56         last 10 log lines:
11:51:56         > + mountOptions=('umask=0077')
11:51:56         > + declare -a mountOptions
11:51:56         > + mountpoint=/boot
11:51:56         > + type=filesystem
11:51:56         > + findmnt /dev/disk/by-partlabel/disk-disk1-ESP /mnt/boot
11:51:56         > + mount /dev/disk/by-partlabel/disk-disk1-ESP /mnt/boot -t vfat -o umask=0077 -o X-mount.mkdir
11:51:56         > + rm -rf /tmp/tmp.Gz19IEgK9U
11:51:56         > warning: the group 'nixbld' specified in 'build-users-group' does not exist
11:51:56         > cp: cannot access '/nix/store/w7hmmw35p0ylsgdy4a33nmrl4qyzii42-duktape-2.7.0': Cannot allocate memory
11:51:56         > [  832.840794] reboot: Power down
11:51:56         For full logs, run 'nix log /nix/store/97963q2q4hxjzi02qsi2j69y2zic2y0x-ghaf-host-disko-images.drv'.
[Pipeline] unstable
11:51:56  WARNING: FAILED: .#packages.x86_64-linux.lenovo-x1-carbon-gen11-debug
[Pipeline] sh
```

gangaram-tii commented:
Rebased to main branch.

brianmcgillion commented on modules/common/systemd/harden.nix:
```diff
@@ -0,0 +1,49 @@
+    ) {} (builtins.attrNames (builtins.readDir configs-dir));
+  };
+in
+  with lib; {
```
remove the use of with

gangaram-tii replied:
Done!
